### PR TITLE
Removed @helper directive

### PR DESCRIPTION
**Pull request by ajaybhargavb**
Issue #281 
@NTaylorMullen

### CODE REVIEW

File: test/Microsoft.AspNet.Razor.Test/Generator/CSharpRazorCodeGeneratorTest.cs [hunk 1]
```diff
@@ -279,22 +273,19 @@ public void CSharpCodeGeneratorCorrectlyGeneratesDesignTimePragmasForUnfinishedE
         [Fact]
         public void CSharpCodeGeneratorCorrectlyGeneratesDesignTimePragmasMarkupAndExpressions()
         {
+            //System.Diagnostics.Debugger.Launch();
```
Comment (dougbu) [inline test/Microsoft.AspNet.Razor.Test/Generator/CSharpRazorCodeGeneratorTest.cs]: remove :smiley_cat:

File: src/Microsoft.AspNet.Razor/Generator/Compiler/CodeBuilder/CSharp/Visitors/CSharpHelperVisitor.cs [hunk 2]
```diff
@@ -1,72 +0,0 @@
...
-        protected override void Visit(HelperChunk chunk)
-        {
-            IDisposable lambdaScope = null;
-
-            var accessibility = "public " + (Context.Host.StaticHelpers ? "static" : String.Empty);
-
-            // We want to write the method signature at 0 indentation so if helper's are formatted they format correctly.
-            var currentIndentation = Writer.CurrentIndent;
-            Writer.ResetIndent();
-            Writer.Write(accessibility).Write(" ").Write(Context.Host.GeneratedClassContext.TemplateTypeName).Write(" ");
```
Comment (dougbu) [inline src/Microsoft.AspNet.Razor/Generator/Compiler/CodeBuilder/CSharp/Visitors/CSharpHelperVisitor.cs]: can't this be removed from `GeneratedClassContext`?

**Comment by dougbu** [comment]
:watch: just one more bit of cleanup

File: src/Microsoft.AspNet.Razor/Generator/Compiler/CodeBuilder/CSharp/Visitors/CSharpHelperVisitor.cs (hunk 2 above)
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Razor/Generator/Compiler/CodeBuilder/CSharp/Visitors/CSharpHelperVisitor.cs]: No he wont be able to.It's also used for things like `@<p></p>`

File: src/Microsoft.AspNet.Razor/Parser/CSharpCodeParser.Directives.cs [hunk 3]
```diff
@@ -24,7 +24,6 @@ private void SetupDirectives()
             MapDirectives(InheritsDirective, SyntaxConstants.CSharp.InheritsKeyword);
             MapDirectives(FunctionsDirective, SyntaxConstants.CSharp.FunctionsKeyword);
             MapDirectives(SectionDirective, SyntaxConstants.CSharp.SectionKeyword);
-            MapDirectives(HelperDirective, SyntaxConstants.CSharp.HelperKeyword);
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Razor/Parser/CSharpCodeParser.Directives.cs]: You should be able to remove the `SyntaxConstants.CSharp.HelperKeyword`

File: src/Microsoft.AspNet.Razor/Parser/CSharpCodeParser.Directives.cs [hunk 4]
```diff
@@ -71,178 +70,6 @@ protected virtual void LayoutDirective()
...
-            Context.CurrentBlock.Type = BlockType.Helper;
-
-            // Verify we're on "helper" and accept
-            AssertDirective(SyntaxConstants.CSharp.HelperKeyword);
-            var block = new Block(CurrentSymbol.Content.ToString().ToLowerInvariant(), CurrentLocation);
-            AcceptAndMoveNext();
-
-            if (nested)
-            {
-                Context.OnError(CurrentLocation, RazorResources.ParseError_Helpers_Cannot_Be_Nested);
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Razor/Parser/CSharpCodeParser.Directives.cs]: Should be able to remove this resource.

File: src/Microsoft.AspNet.Razor/Parser/CSharpCodeParser.Directives.cs [hunk 5]
```diff
@@ -71,178 +70,6 @@ protected virtual void LayoutDirective()
...
-                    error = RazorResources.ErrorComponent_EndOfFile;
-                }
-                else
-                {
-                    error = RazorResources.FormatErrorComponent_Character(CurrentSymbol.Content);
-                }
-
-                Context.OnError(
-                    CurrentLocation,
-                    RazorResources.FormatParseError_Unexpected_Character_At_Helper_Name_Start(error));
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Razor/Parser/CSharpCodeParser.Directives.cs]: Should be able to remove resource

File: src/Microsoft.AspNet.Razor/Parser/CSharpCodeParser.Directives.cs [hunk 6]
```diff
@@ -71,178 +70,6 @@ protected virtual void LayoutDirective()
...
-            // Expecting parameter list start: "("
-            var bracketErrorPos = CurrentLocation;
-            if (!Optional(CSharpSymbolType.LeftParenthesis))
-            {
-                if (!errorReported)
-                {
-                    errorReported = true;
-                    Context.OnError(
-                        CurrentLocation,
-                        RazorResources.FormatParseError_MissingCharAfterHelperName("("));
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Razor/Parser/CSharpCodeParser.Directives.cs]: ditto

File: src/Microsoft.AspNet.Razor/Parser/CSharpCodeParser.Directives.cs [hunk 7]
```diff
@@ -71,178 +70,6 @@ protected virtual void LayoutDirective()
...
-                var bracketStart = CurrentLocation;
-                if (!Balance(BalancingModes.NoErrorOnFailure,
-                             CSharpSymbolType.LeftParenthesis,
-                             CSharpSymbolType.RightParenthesis,
-                             bracketStart))
-                {
-                    errorReported = true;
-                    Context.OnError(
-                        bracketErrorPos,
-                        RazorResources.ParseError_UnterminatedHelperParameterList);
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Razor/Parser/CSharpCodeParser.Directives.cs]: ditto

File: test/Microsoft.AspNet.Razor.Test/TestFiles/CodeGenerator/CS/Source/InlineBlocks.cshtml [hunk 8]
```diff
@@ -1,3 +0,0 @@
-﻿@helper Link(string link) {
```
Comment (NTaylorMullen) [inline test/Microsoft.AspNet.Razor.Test/TestFiles/CodeGenerator/CS/Source/InlineBlocks.cshtml]: Don't remove this file, just change from `@helper` => `@section`

File: test/Microsoft.AspNet.Razor.Test/Generator/CSharpRazorCodeGeneratorTest.cs [hunk 9]
```diff
@@ -75,12 +75,6 @@ public void ConstructorRequiresNonNullHost()
         [InlineData("Templates")]
         [InlineData("Sections")]
         [InlineData("RazorComments")]
-        [InlineData("Helpers")]
-        [InlineData("HelpersMissingCloseParen")]
-        [InlineData("HelpersMissingOpenBrace")]
-        [InlineData("HelpersMissingOpenParen")]
-        [InlineData("NestedHelpers")]
-        [InlineData("InlineBlocks")]
```
Comment (NTaylorMullen) [inline test/Microsoft.AspNet.Razor.Test/Generator/CSharpRazorCodeGeneratorTest.cs]: undo this specific one (left comment on file)

File: test/Microsoft.AspNet.Razor.Test/TestFiles/CodeGenerator/CS/Output/InlineBlocks.cs [hunk 10]
```diff
@@ -1,84 +0,0 @@
-#pragma checksum "InlineBlocks.cshtml" "{ff1816ec-aa5e-4d10-87f7-6f4963833460}" "85fc1bd0306a5a6164d3d866bd690ff95cba0a8e"
```
Comment (NTaylorMullen) [inline test/Microsoft.AspNet.Razor.Test/TestFiles/CodeGenerator/CS/Output/InlineBlocks.cs]: Don't remove, see comments on `cshtml` equivalent

File: test/Microsoft.AspNet.Razor.Test/Generator/CSharpTagHelperRenderingTest.cs [hunk 11]
```diff
@@ -388,7 +388,6 @@ public void TagHelpers_Directive_GenerateDesignTimeMappings()
 
         [Theory]
         [InlineData("TagHelpersInSection")]
```
Comment (NTaylorMullen) [inline test/Microsoft.AspNet.Razor.Test/Generator/CSharpTagHelperRenderingTest.cs]: Modify this test to just validate the section now, no need for a theory. Update test name accordingly.

File: test/Microsoft.AspNet.Razor.Test/Parser/CSharp/CSharpDirectivesTest.cs [hunk 12]
```diff
@@ -389,25 +389,5 @@ public void SectionDirective()
                     Factory.MetaCode("}")
                            .Accepts(AcceptedCharacters.None)));
         }
-
-        [Fact]
-        public void HelperDirective()
-        {
-            ParseBlockTest("@helper Strong(string value) { foo(); }",
-                new HelperBlock(new HelperCodeGenerator(new LocationTagged<string>("Strong(string value) {", new SourceLocation(8, 0, 8)), headerComplete: true),
```
Comment (NTaylorMullen) [inline test/Microsoft.AspNet.Razor.Test/Parser/CSharp/CSharpDirectivesTest.cs]: Should be able to remove the `HelperBlock`

File: test/Microsoft.AspNet.Razor.Test/TestFiles/CodeGenerator/CS/Source/Instrumented.cshtml [hunk 13]
```diff
@@ -1,8 +1,4 @@
-﻿@helper Strong(string s) {
```
Comment (NTaylorMullen) [inline test/Microsoft.AspNet.Razor.Test/TestFiles/CodeGenerator/CS/Source/Instrumented.cshtml]: Change this to be `@if (true) {....}`

**Comment by NTaylorMullen** [comment]
:watch:

File: test/Microsoft.AspNet.Razor.Test/TestFiles/CodeGenerator/CS/Source/Instrumented.cshtml (hunk 13 above)
Comment (NTaylorMullen) [inline test/Microsoft.AspNet.Razor.Test/TestFiles/CodeGenerator/CS/Source/Instrumented.cshtml]: Ignore this comment.

File: src/Microsoft.AspNet.Razor/Parser/CSharpCodeParser.Directives.cs (hunk 3 above)
Comment (ajaybhargavb) [inline src/Microsoft.AspNet.Razor/Parser/CSharpCodeParser.Directives.cs]: removed.

**Comment by ajaybhargavb** [comment]
Updated.

**Comment by dougbu** [comment]
:shipit: once @NTaylorMullen is moderately pleased

**Comment by NTaylorMullen** [comment]
Ensure this doesn't break anything in MVC then :shipit:, if it does we can do a quick react PR.

**Pull request merged**